iso15118/shared/security.py

```diff
@@ -1125,6 +1130,218 @@ def decrypt_priv_key(
     raise DecryptionError()
 
 
+def derive_certificate_hash_data(
+    certificate: bytes, issuer_certificate: bytes
+) -> Dict[str, str]:
+    """Extract certificate hash data to be used in an OCPP AuthorizeRequest.
+
+    Args:
+        certificate: A certificate in binary (DER) form.
+        issuer_certificate: The certificate used for signing `certificate`,
+            in binary (DER) form.
+            For a self-signed certificate, these will be the same.
+
+    Returns:
+        A dictionary with all information required for an OCSPRequestDataType
+        (2.36. OCSPRequestDataType, p. 382, OCPP 2.0.1 Part 2)
+
+    Raises:
+        CertAttributeError: if a certificate is provided with a hash algorithm
+            that OCPP doesn't accept.
+            Only SHA256, SHA384, and SHA512 are allowed.
+            (3.42 HashAlgorithmEnumType, p. 403, OCPP 2.0.1 Part 2)
+    """
+    certificate = load_der_x509_certificate(certificate)
+    issuer_certificate = load_der_x509_certificate(issuer_certificate)
+    builder = OCSPRequestBuilder().add_certificate(
+        certificate, issuer_certificate, certificate.signature_hash_algorithm
+    )
+
+    ocsp_request = builder.build()
+
+    # For the hash algorithm, convert to the naming used in OCPP.
+    # Only SHA256, SHA384, and SHA512 are allowed in OCPP 2.0.1.
+    hash_algorithm_for_ocpp = certificate.signature_hash_algorithm.name.upper()
+    if hash_algorithm_for_ocpp not in {"SHA256", "SHA384", "SHA512"}:
+        raise CertAttributeError("Unknown hash algorithm")
+
+    try:
+        responder_url = get_ocsp_url_for_certificate(certificate)
+    except (ExtensionNotFound, OCSPServerNotFoundError):
+        # TODO GitHub#96: This may just result in failure down the road.
+        # Should we let this fail on these exceptions, or is there
+        # another way to try to get a responder_url?
+        responder_url = "https://www.example.com/"
```

Comment:
if we cant get the OCSP url, then the OCSP call cant happen, so we may just drop everything and end as a failure, ie authorization has failed

Reply:
This might be the best way we have to resolve #96 then.